File: CodeForces/ECR167/D/main.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>
#include <climits>

using namespace std;

#define INF INT_MAX
#define MOD 998244353

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, m;
    cin >> n >> m;

    vector<int> a(n), b(n), c(m);
    for (int i = 0; i < n; ++i)
        cin >> a[i];
    for (int i = 0; i < n; ++i)
        cin >> b[i];
    for (int i = 0; i < m; ++i)
        cin >> c[i];

    vector<tuple<int, int, int>> ab(n);
    for (int i = 0; i < n; ++i)
    {
        ab[i] = make_tuple(a[i], b[i], a[i] - b[i]);
    }

    sort(ab.begin(), ab.end());

    vector<tuple<int, int, int>> ab_new;
    int tmp = INF;
    for (auto &[i, j, k] : ab)
    {
        if (tmp > k)
        {
            tmp = k;
            ab_new.emplace_back(i, j, k);
        }
    }

    ab = ab_new;

    vector<pair<int, int>> score;
    for (int idx = 0; idx < ab.size(); ++idx)
    {
        score.emplace_back(get<2>(ab[idx]), idx);
    }

    vector<int> a_sorted;
    for (auto &[i, j, k] : ab)
    {
        a_sorted.push_back(i);
    }

    vector<pair<int, int>> min_;
    for (auto &s : score)
    {
        if (min_.empty())
        {
            min_.push_back(s);
        }
        else
        {
            min_.push_back(min(min_.back(), s));
        }
    }

    long long ans = 0;
    for (int i : c)
    {
        int amari = i;
        while (amari > 0)
        {
            auto it = upper_bound(a_sorted.begin(), a_sorted.end(), amari);
            if (it == a_sorted.begin())
            {
                break;
            }
            int idx = distance(a_sorted.begin(), it) - 1;
            auto [min_value, min_idx] = min_[idx];
            int tmp = (amari - a_sorted[min_idx]) / min_value + 1;
            ans += tmp * 2;
            amari -= tmp * min_value;
        }
    }

    cout << ans << endl;

    return 0;
}

File: CodeForces/ECR167/D/main.py
import sys
from bisect import bisect_left, bisect_right

import string

# 外部ライブラリ
# from sortedcontainers import SortedSet, SortedList, SortedDict

alph_s = tuple(string.ascii_lowercase)
alph_l = tuple(string.ascii_uppercase)
around4 = ((-1, 0), (1, 0), (0, -1), (0, 1))  # 上下左右
around8 = ((-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1))
inf = float("inf")
mod = 998244353
input = lambda: sys.stdin.readline().rstrip()
pritn = lambda *x: print(*x)
PY = lambda: print("Yes")
PN = lambda: print("No")
SI = lambda: input()
IS = lambda: input().split()
II = lambda: int(input())
MII = lambda: map(int, input().split())
LMII = lambda: list(map(int, input().split()))


n, m = MII()

a = LMII()
b = LMII()
c = LMII()
ans = 0
# 作れるやつの中でコスパが最も良いものを選びたい。

ab = [(i, j, i - j) for i, j in zip(a, b)]

ab.sort(key=lambda x: x[0])
ab_new = []
tmp = inf
for i, j, k in ab:
    if tmp > k:
        tmp = k
        ab_new.append((i, j, k))

ab = ab_new

score = [(i[2], idx) for idx, i in enumerate(ab)]
a = [i[0] for i in ab]
min_ = []

for i in score:
    if not min_:
        min_.append(i)
    else:
        min_.append(min(min_[-1], i))

for i in c:
    amari = i
    while amari > 0:
        idx = bisect_right(a, amari)
        if idx == 0:
            break
        rslt = min_[idx - 1]
        tmp = (amari - a[rslt[1]]) // rslt[0] + 1
        ans += tmp * 2
        amari -= tmp * rslt[0]

print(ans)
